java/fury-core/src/main/java/org/apache/fury/Fury.java

```diff
@@ -1221,15 +1240,55 @@ public Object deserializeJavaObjectAndClass(FuryReadableChannel channel) {
     return deserializeJavaObjectAndClass(buf);
   }
 
+  @SuppressWarnings("unchecked")
   @Override
   public <T> T copy(T obj) {
     if (Objects.isNull(obj)) {
       return null;
     }
-    if (obj instanceof FuryCopyable) {
-      return (T) ((FuryCopyable) obj).copy(this);
-    } else {
-      return (T) classResolver.getOrUpdateClassInfo(obj.getClass()).getSerializer().copy(obj);
+    try {
+      copyDepth++;
+      if (Objects.isNull(originToCopyMap)) {
+        originToCopyMap = new IdentityMap();
+      }
+      Object newObj = originToCopyMap.get(obj);
+      if (Objects.nonNull(newObj)) {
+        return (T) newObj;
+      }
+      if (copyRefTracking) {
+        originObjRef = obj;
+      }
+      T copy;
+      if (obj instanceof FuryCopyable) {
+        copy = (T) ((FuryCopyable<?>) obj).copy(this);
+      } else {
+        copy = (T) classResolver.getOrUpdateClassInfo(obj.getClass()).getSerializer().copy(obj);
+      }
+      if (Objects.nonNull(originObjRef)) {
+        copyReference(copy);
+      }
+      return copy;
+    } catch (StackOverflowError e) {
+      throw processCopyStackOverflowError(e);
+    } finally {
+      copyDepth--;
+      if (copyDepth == 0) {
+        resetCopy();
+      }
+    }
+  }
+
+  /**
+   * Record the mapping between the object currently being copied and the newly generated object
+   * after copying.
+   *
+   * @param obj the new object instance
+   */
+  @SuppressWarnings("unchecked")
+  public void copyReference(Object obj) {
+    if (copyDepth > 0 && Objects.nonNull(originObjRef) && Objects.nonNull(obj)) {
+      originToCopyMap.put(originObjRef, obj);
```

Comment:
Could we add a new method in `RefResolver` and reuse the IndentifyMap in this ref resolver?

Reply:
Okay, I'll refer to the circular dependency processing of serialize/deserialize and try to implement it based on MapRefResolver